Second deliverable

FernandoGal25 opened the pull request:
* Fixed from 1st deliverable
* Changed domain entity Item -> Pokemon
* Creates rest repository to request from external api
* Creates new create pokemon use case
* Updates csv repository to persist new records

Commit:
Fixes of GET endpoint
- Creates new GET endpoint that return all items
- Specifies domain entity being Pokemon
- Wraps errors and handles each to generate readable responses
- Addes code comments
- Generates unit tests for each endpoint
- Creates a pokemon mock repository that helps in the unit testing

Commit:
Starts contruction of POST endpoint
- Creates rest repository to request from external api
- Creates new create pokemon use case
- Updates Mcsv repository to persist new records

RolloCasanova commented on application/contracts/pokemon_repository.go:
```diff
@@ -0,0 +1,13 @@
+	Interface of pokemon repository, defines the methods of the repository needed to handle
+	the entity in question.
+*/
+type PokemonRepository interface {
```
Exported elements should be documented with a brief description using `//` comments (as part of Go's best practices - https://go.dev/blog/godoc); please change this and all other occurrences of this specific issue all over the project.

```suggestion
// PokemonRepositare interface defines the methods of the repository needed to handle.
type PokemonRepository interface {
```

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,68 @@
+	errors "github.com/FernandoGal25/academy-go-q42021/error"
+)
+
+const TOTAL_POKEMON = 898
```
Constants should not be written in `SNAKE_CAPS_CASE`, only `camelCase` and `PascaCase` are allowed.

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,68 @@
+			Message: "OUT OF BOUND, CURRENTLY THERE ARE ONLY " + strconv.Itoa(TOTAL_POKEMON) + " POKEMON",
+		}
+	}
+	p, err := s.CSVPokemonRepository.FindByID(key)
```
Assignments should be only cuddled with other assignments; it gives better readability to your code and separates distinct functionality code blocks. Take it in mind for future deliveries.

```suggestion
	}
	
	p, err := s.CSVPokemonRepository.FindByID(key)
```

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,68 @@
+func (s PokemonService) GetPokemonByID(key uint64) (*model.Pokemon, error) {
+	if key > TOTAL_POKEMON {
+		return nil, errors.DomainValidationError{
+			Message: "OUT OF BOUND, CURRENTLY THERE ARE ONLY " + strconv.Itoa(TOTAL_POKEMON) + " POKEMON",
```
It seems you are missing the Err field of the DomainValidationError, or is it intended?

You could also use `fmt.Sprintf()` method to format your message and avoid having conversions of values.

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	p, err := s.CSVPokemonRepository.FindByID(key)
+	if err != nil {
+		return nil, err
```
Errors returned from interfaces should be wrapped. This is useful in the future when there's some error, and by simply checking the error message you could get an idea of what's happening.

Please fix all other occurrences of this issue all over your project.

RolloCasanova commented on application/usecase/pokemon_test.go:
```diff
@@ -0,0 +1,44 @@
+package usecase
```
Refer to https://dave.cheney.net/2019/05/07/prefer-table-driven-tests, it helps to develop test case scenarios in a more organized way.

Good use of `gomock` by the way.

RolloCasanova commented on config/config.go:
```diff
@@ -20,9 +19,11 @@ type config struct {
+	Extracts the config variables.
+*/
+func ReadConfig() (*config, error) {
+	var C config
```
No need to have `C` exported, you are only using it in this function.

Also, there's no need to declare two variables (line 26 and 27), you could simple initialize it of type *config, and that's it.

```suggestion
	var c *config
```

[EXTRA] You should never ignore errors (line 31), and remember to wrap errors from external packages when also returning an error.

RolloCasanova commented on config/config.go:
```diff
@@ -20,9 +19,11 @@ type config struct {
+/*
+	Extracts the config variables.
+*/
+func ReadConfig() (*config, error) {
```
Could be annoying to export an unexported type.

RolloCasanova commented on domain/model/pokemon.go:
```diff
@@ -0,0 +1,13 @@
+	Height         int    `json:"height"`
+	Weight         int    `json:"weight"`
+	Order          int    `json:"order"`
+	BaseExperience int    `json:"base_experience"`
```
JSON standard is camelCase.
```suggestion
	BaseExperience int    `json:"baseExperience"`
```

RolloCasanova commented on error/errors.go:
```diff
@@ -0,0 +1,78 @@
+/*
+	Custom message made in case of undefined error.
+*/
+const DEFAULT_MESSAGE = "SOMETHING WENT TERRIBLY WRONG"
```
Constants are not written in `CAPS_SNAKE_CASE`, better use `camelCase` or `PascalCase`.

Also, the message should not be written in `ALL CAPS`, better be written in `Sentence case`,

RolloCasanova commented on error/errors.go:
```diff
@@ -0,0 +1,78 @@
+*/
+const DEFAULT_MESSAGE = "SOMETHING WENT TERRIBLY WRONG"
+
+/*
+	Custom error made for wrapping CSV handling errors.
+*/
+type CSVFormatError struct {
+	Message string
+	Err     error
+}
+
+/*
+	Custom error made in case the searched item is not found.
+*/
+type EntityNotFoundError struct {
+	Message string
+	Err     error
+}
+
+/*
+	Custom error that wraps the errors made on request.
+*/
+type InvalidRequestError struct {
+	Message string
+	Err     error
+}
+
+/*
+	Custom error made in case any business rule is not met.
+*/
+type DomainValidationError struct {
+	Message string
+	Err     error
+}
```
Instead of `FooBarError`, errors should be in the form `ErrFooBar`.

Remember to change the documentation with the `//` form.

RolloCasanova commented on infrastructure/datastore/csv.go:
```diff
@@ -6,34 +6,90 @@ import (
+*/
 func (h *CSVHandler) BuildHandler() error {
-	f, err := os.Open(h.Path)
+	wf, err := os.OpenFile(h.path, os.O_WRONLY|os.O_APPEND, 0666)
+
+	if err != nil {
```
```suggestion
	wf, err := os.OpenFile(h.path, os.O_WRONLY|os.O_APPEND, 0666)
	if err != nil {
```

RolloCasanova commented on infrastructure/datastore/csv.go:
```diff
@@ -6,34 +6,90 @@ import (
 
+/*
+	Wrapper of the CSV file, combines the management
+	of the os and enconding/csv libraries.
```
```suggestion
	of the os and encoding/csv libraries.
```

RolloCasanova commented on infrastructure/datastore/csv.go:
```diff
@@ -6,34 +6,90 @@ import (
+		return errors.New("CANNOT READ CSV FILE")
+	}
 
+	h.schema = header
```
```suggestion
	h.schema = header
	
```

RolloCasanova commented on infrastructure/datastore/csv.go:
```diff
@@ -6,34 +6,90 @@ import (
+	wf, err := os.OpenFile(h.path, os.O_WRONLY|os.O_APPEND, 0666)
+
+	if err != nil {
+		return errors.New("CANNOT OPEN CSV FILE")
```
Do not dynamically create errors, create them in the errors package and use them here. `Sentence case` and wrap errors before returning them.

Fix all other occurrences.

RolloCasanova commented on infrastructure/router/router.go:
```diff
@@ -1,18 +1,23 @@
 
-	e.GET("/items/:id", func(context echo.Context) error { return c.Item.GetItemById(context) })
-	// e.POST("/items", func(context echo.Context) error { return c.Item.CreateItem(context) })
+	e.GET("/pokemon/:ID", func(context echo.Context) error { return c.Pokemon.ActionGetByID(context) })
+	e.GET("/pokemon", func(context echo.Context) error { return c.Pokemon.ActionGetAll(context) })
+	e.POST("/pokemon/:ID", func(context echo.Context) error { return c.Pokemon.ActionPostByID(context) })
```
Use plural for resource in path, and it's a good idea to have them separated by endpoint. Also, it's not common to use CAPS for query params.

```suggestion
	e.GET("/pokemons", func(context echo.Context) error { return c.Pokemon.ActionGetAll(context) })
	
	e.GET("/pokemons/:id", func(context echo.Context) error { return c.Pokemon.ActionGetByID(context) })
	e.POST("/pokemons/:id", func(context echo.Context) error { return c.Pokemon.ActionPostByID(context) })
```

RolloCasanova commented on interface/controller/app_controller.go:
```diff
@@ -1,5 +1,73 @@
 package controller
```
Is redundant to name the file app_controller, as you are inside the controller package. Please consider fixing this naming redundancy here and all other files in your project.

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+
+	if i.ID == 0 {
+		return nil, customErrors.EntityNotFoundError{
+			Message: "THE SEARCHED POKEMON HAS NOT BEEN REGISTERED IN THIS DATABASE",
```
Missing Err from EntityNotFoundError This would be a good place to use a custom error denoting there's no pokemon with such ID.

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+
+	for {
+		record, err := r.Handler.Read()
+		if err == io.EOF {
```
Better use `error.Is() method` to check for specific errors. In wrapped errors, comparing with `==` will result in giving a false negative.

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+*/
+func (r CSVPokemonRepository) FindByID(ID uint64) (*model.Pokemon, error) {
+	if err := r.Handler.BuildHandler(); err != nil {
+		return nil, err
```
Wrap external package errors before returning them. Check and fix all other occurrences of this specific issue.

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+
+	h, err := strconv.Atoi(data[2])
+	if err != nil {
+		return customErrors.CSVFormatError{Message: "CSV ITEM DOES NOT POSSESS A VALID HEIGHT", Err: err}
```
Could initialize a variable with this message, and use `fmt.Sprintf` to avoid repeating yourself.

``` go
const CSVItemMessage = "CVS item does not posses a valid %s attribute"
...
return customErrors.CSVFormatError{Message: fmt.Sprintf(CSVItemMessage, "heigth"), Err: err}
return customErrors.CSVFormatError{Message: fmt.Sprintf(CSVItemMessage, "weight"), Err: err}
return customErrors.CSVFormatError{Message: fmt.Sprintf(CSVItemMessage, "order"), Err: err}
return customErrors.CSVFormatError{Message: fmt.Sprintf(CSVItemMessage, "baseExperience"), Err: err}

// Or even better, generate a function in customErrors accepting both parameters, and return an instance of this error.

func GenerateCSVFormatErr(attribute string, err error) CSVFormatError {
	return customErrors.CSVFormatError{Message: fmt.Sprintf(CSVItemMessage, atribute), Err: err}
}

return customErrors.GenerateCSVFormatErr("height", err)
return customErrors.GenerateCSVFormatErr("weight", err)
return customErrors.GenerateCSVFormatErr("order", err)
return customErrors.GenerateCSVFormatErr("baseExperience", err)
```

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+		parsedID, err := strconv.ParseUint(record[0], 10, 16)
+		if err != nil {
+			return nil, customErrors.CSVFormatError{Message: "CSV ITEM DOES NOT POSSESS A VALID ID", Err: err}
+		}
+		if parsedID == ID {
```
```suggestion
		}
		
		if parsedID == ID {
```

RolloCasanova commented on interface/repository/csv_pokemon_repository.go:
```diff
@@ -0,0 +1,148 @@
+		if err != nil {
+			return nil, customErrors.CSVFormatError{Message: "CSV ITEM DOES NOT POSSESS A VALID ID", Err: err}
+		}
+		var i model.Pokemon
```
`i` is not a good name for this variable.

RolloCasanova commented on interface/repository/rest_pokemon_repository.go:
```diff
@@ -0,0 +1,65 @@
+	return RESTPokemonRepository{p}
+}
+
+func (r RESTPokemonRepository) FindByID(ID uint64) (*model.Pokemon, error) {
```
Just for curiosity, why use uint64, if you will be converting that value to int?

RolloCasanova commented on interface/repository/rest_pokemon_repository.go:
```diff
@@ -0,0 +1,65 @@
+}
+
+func (r RESTPokemonRepository) FindByID(ID uint64) (*model.Pokemon, error) {
+	resp, err := http.Get(r.APIGateway + "/pokemon/" + strconv.FormatUint(ID, 10))
```
Could use `fmt.Sprintf` to format and generate a string.

RolloCasanova commented on interface/repository/rest_pokemon_repository.go:
```diff
@@ -0,0 +1,65 @@
+func (r RESTPokemonRepository) FindByID(ID uint64) (*model.Pokemon, error) {
+	resp, err := http.Get(r.APIGateway + "/pokemon/" + strconv.FormatUint(ID, 10))
+	if err != nil {
+		return nil, errors.New("REQUEST ERROR")
```
Wrap errors before returning them. No ALL CAPS. Fix for this and all occurrences.

RolloCasanova reviewed (changes requested):
A few things to improve:

- Documentation of your code, refer to https://go.dev/blog/godoc to know how to do so.
- Format of some variable, go only admits `camelCase` and `PascalCase`, careful with that.
- For unit tests refer to https://dave.cheney.net/2019/05/07/prefer-table-driven-tests, and also add tests for the controller, router, and repository package's files.
- Error handling when an invalid id is passed in the external API endpoint. The app should exit when a non-int id is added as a value. You should handle that error by responding with an appropriate HTTP status and letting the end-user keep using the app.

RolloCasanova commented:
Forgot to mention:
DO NOT DELETE THIS PR, add changes and corrections in the same branch, and after you finish send me a DM so I can review the latest changes.

Good job!

Commit:
Corrections of write endpoint
- Changes name of files to avoid redundancy eg.: controller/pokemon_controller.go => controller/pokemon.go
- Wraps errors in multiple wrappers in order to track the source
- Adds case use unit testing
- CCorrects documentation of ecode

Commit:
Adds error wrappers in rest pokemon repository

Commit:
Adds default config.yml

RolloCasanova commented on application/contracts/pokemon_repository.go:
```diff
@@ -0,0 +1,11 @@
+
+import "github.com/FernandoGal25/academy-go-q42021/domain/model"
+
+// Interface of pokemon repository, defines the methods of the repository
```
Documentation comments should start with the name of the element. Please fix this for all other occurrences of this issue.
```suggestion
// PokemonRepository interface of pokemon repository, defines the methods of the repository
```

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,82 @@
+		return nil, err
+	}
+
+	p, err := s.CSVPokemonRepository.FindByID(key)
+
+	if err != nil {
+		return nil, errors.ErrRepositoryWrapper{Message: "Failed to fetch pokemon from repository", Err: err}
+	}
```
Keep together error assignation with error check in a single code block.

```suggestion
	p, err := s.CSVPokemonRepository.FindByID(key)
	if err != nil {
		return nil, errors.ErrRepositoryWrapper{Message: "Failed to fetch pokemon from repository", Err: err}
	}
```

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,82 @@
+
+// Orchestrates the tools required to retrieve all pokemon.
+func (s PokemonService) GetAllPokemons() ([]model.Pokemon, error) {
+	p, err := s.CSVPokemonRepository.FetchAll()
+
+	if err != nil {
+		return nil, errors.ErrRepositoryWrapper{Message: "Failed to fetch pokemons from repository", Err: err}
+	}
```
```suggestion
	p, err := s.CSVPokemonRepository.FetchAll()
	if err != nil {
		return nil, errors.ErrRepositoryWrapper{Message: "Failed to fetch pokemons from repository", Err: err}
	}
```

RolloCasanova commented on application/usecase/pokemon.go:
```diff
@@ -0,0 +1,82 @@
+func validateRangeID(key int) error {
+	if key > totalPokemon {
+		return errors.ErrDomainValidation{
+			Message: fmt.Sprintf("Out of bound, pokemon with ID: %v searched while there are currently only %v pokemon.", key, totalPokemon),
```
Even this is not using an already generated error you could initialize one and make use here. It's always a good idea to be consistent in returning an error object so it can be wrapped by the previous layer(s).

Could be an error defined in error package, or at the same usecase level.
```suggestion
			Message: fmt.Sprintf("Out of bound, pokemon with ID: %v searched while there are currently only %v pokemon.", key, totalPokemon),
			Err: pokemonError.ErrOutOfBounds,
```

RolloCasanova commented on application/usecase/pokemon_test.go:
```diff
@@ -0,0 +1,256 @@
+}
+
+func TestPokemonService_GetPokemonByID(t *testing.T) {
+
```
```suggestion
```

RolloCasanova commented on application/usecase/pokemon_test.go:
```diff
@@ -0,0 +1,256 @@
+	repo2 := repository.NewMockPokemonRepository(mockCtrl)
+	service := PokemonService{CSVPokemonRepository: repo2, RestPokemonRepository: repo1}
+	for _, tt := range tests {
+
```
```suggestion
```

RolloCasanova commented on config/config.go:
```diff
@@ -20,26 +20,28 @@ type config struct {
 	viper.SetConfigName("config")
 	viper.SetConfigType("yml")
-	path, _ := os.Getwd()
+	path, err := os.Getwd()
+
+	if err != nil {
```
```suggestion
	path, err := os.Getwd()
	if err != nil {
```

RolloCasanova commented on error/errors.go:
```diff
@@ -0,0 +1,132 @@
+package errors
```
The content of this file is repeated. In places like this is where an interface becomes handy to avoid these kinds of scenarios.

Having already defined the error messages as constants (either the full message, or a formatted message), is also a good idea.

``` golang
package errors

import "fmt"

// ErrorInterface is an interface that all errors must implement.
type ErrorInterface interface {
	Error() string
	Unwrap() error
}

// PokeError struct is used to represent an error.
type PokeError struct {
	Message string
	Err     error
}

// PokeErrors variable definitions.
var (
	ErrSample1 ErrorInterface = PokeError{Message: "Sample 1 error message"}
	ErrSample2 ErrorInterface = PokeError{Message: "Sample 2 error message"}
	ErrSample3 ErrorInterface = PokeError{Message: "Sample 3 error message"}
)

// PokeErrorf creates a new PokeError with the given message.
func PokeErrorf(format string, args ...interface{}) PokeError {
	return PokeError{Message: fmt.Sprintf(format, args...)}
}

// New returns a new error with the given message.
func New(message string, err error) *PokeError {
	return &PokeError{Message: message, Err: err}
}

// ReplaceMessage replaces the error message.
func (p *PokeError) ReplaceMessage(message string) *PokeError {
	p.Message = message

	return p
}

// AppendError appends the given error to the current PokeError.
func (p *PokeError) ReplaceError(err error) *PokeError {
	p.Err = err

	return p
}

// Error returns the error message.
func (p PokeError) Error() string {
	return p.Message
}

// Unwrap returns the underlying error.
func (p PokeError) Unwrap() error {
	return p.Err
}

```

RolloCasanova commented on helpers/helpers.go:
```diff
@@ -0,0 +1,28 @@
+// Transform snake to camel case, function copied from
+// https://www.socketloop.com/tutorials/golang-underscore-or-snake-case-to-camel-case-example
+func SnakeCaseToCamelCase(snake string) (camel string) {
+
```
```suggestion
```

RolloCasanova commented on helpers/helpers.go:
```diff
@@ -0,0 +1,28 @@
+				}
+			}
+		}
+	}
+	return
```
```suggestion
	}
	
	return
```

RolloCasanova commented on interface/controller/app.go:
```diff
@@ -0,0 +1,77 @@
+		}
+	}
+
+	if status == 0 {
+		status = http.StatusInternalServerError
+	}
```
Instead of declaring `status` in line 42, you can initialize it with this error and avoid checking `status == 0`. In the end, you just use whatever value `status` has.

RolloCasanova reviewed (approved):
Still missing unit tests on other layers (controller, usecase, router).

Accept interfaces when initializing a new instance of these layers. This will help you to create these unit tests by using a mock instead of a concrete type.

I saw you changed the comments to `//` for documentation purposes, you are only missing the name of the component to be the first word of these comments.

Other than that, you did a great job; consider this PR as closed and take these comments to improve the last deliverable.
